Developer Journey - Level 1 docs (#1786)
dfinity/portal: docs/tutorials/developer-journey/level-1/1.3-first-dapp.md

# 1.3: Developing your first dapp

## Overview

At this point in our developer journey, we've covered the fundamentals of canister development and deployed the default 'Hello, world' project template on the Motoko playground. It's time to get into writing code and developing our own first dapp! 

> For this tutorial, we'll be creating a poll dapp that includes both a backend canister and a frontend canister. To recap, backend canisters are used to host the dapp's source code, while frontend canisters are used to host user interface assets, such as HTML and CSS. 

Reviewer: ```suggestion
For this tutorial, we'll be creating a poll dapp that includes both a backend canister and a frontend canister. To recap, backend canisters are used to host the dapp's smart contract code, while frontend canisters are used to host user interface assets, such as HTML and CSS. 
```

In your `src/poll_backend/main.mo` file, paste the following code:

:::caution
The following example is a **code snippet** that is part of a larger code file. This snippet may return an error if run on its own. To view the full code file that should be run, please see [final code](#final-code).
:::

```motoko
actor {

    //actor code goes here
    
}
```

This code defines an actor, but it is currently empty. It does not define any data or send and receive messages. 

Next, let's add some code inside the actor that proposes a question for our poll app. For example, we can ask the user, "What is your favorite programming language?". 

To add this question, insert the following code inside of the actor definition:

:::caution
The following example is a **code snippet** that is part of a larger code file. This snippet may return an error if run on its own. To view the full code file that should be run, please see [final code](#final-code).
:::

```motoko
    var question: Text = "What is your favorite programming language?";
```

:::info

> #### What does this code do?

Reviewer: it looks like having the header inside the admonition breaks the formatting of the admonition. see screenshot.
<img width="636" alt="Screenshot 2023-08-24 at 4 29 29 PM" src="https://github.com/dfinity/portal/assets/98767015/3d08f7b0-1ff6-4b60-8f27-74f77e670514">

- All methods that return values on the IC have to be declared as `async` in order to allow asynchronous execution. 
- Since the `getQuestion` method doesn't change any data, it can be declared as a `query` call. We'll dive further into this shortly. 
- This method uses the `Text` type that we previously have already imported at the top of the file. 
:::

### Query calls vs. update calls

A **query call** is executed on one node of a subnet. It is not possible to use query calls to alter data.

An **update call** is executed on all nodes of a subnet. The result must pass through consensus on the subnet and has the ability to alter data. 

A detailed comparison breakdown can be found below:

|                      | Queries                    | Updates                |
|----------------------|----------------------------|------------------------|
| Resource consumption | Low                        | High                   |
| Response times       | Fast (300ms-900ms)         | Slow (2s-10s)          |

> | Cost                 | Free                       | Cost cycles            |

Reviewer: This table is nice. Since queries calls may incur a cost in the future, should we indicate the cost as "Low" instead of "Free"?

Reviewer 2: Or "(Currently) free".

Do we also want to add a row on "Security"
|                                      | Queries                                        | Updates                                   |
|----------------------|--------------------------------|------------------------------|
| Security                      |  Results not certified by default | Results certified by default    |

| Data change         | Not allowed to change data | Allowed to change data |

After adding the `getQuestion` method, your `main.mo` file should look like this:

```motoko
import Text "mo:base/Text";

actor {
    var question: Text = "What is your favorite programming language?";

    public query func getQuestion() : async Text { 
      question 
    };
}
```

### Creating a data structure to store the data

In our poll, we will be storing each poll option and a number associated with how many votes each option has. This collection of data will be in the form of a *key-value store*. In other languages, this data structure is known as a "dictionary" or a "map". In Motoko, this data structure is called an [`RBTree`](/motoko/main/base/RBTree.md). 

> `RBTreee` maps a **vote option** of type `Text` to a **vote count** of type `Nat`. `Text` is a data type for string text values, and `Nat` is a data type for natural numbers, which are whole numbers without decimal points. 

Reviewer 2: ```suggestion
`RBTree` maps a **vote option** of type `Text` to a **vote count** of type `Nat`. `Text` is a data type for string text values, and `Nat` is a data type for natural numbers, which are whole numbers without decimal points. 
```

//      * [["Motoko","1"],["Python","0"],["Rust","0"],["TypeScript","0"]]
    
  public func vote(entry: Text) : async [(Text, Nat)] {

    //Check if the entry already has votes.
    //Note that "votes_for_entry" is of type ?Nat. This is because: 
    // * If the entry is in the RBTree, the RBTree returns a number.
    // * If the entry is not in the RBTree, the RBTree returns `null` for the new entry.
    let votes_for_entry :?Nat = votes.get(entry);
    
    //Need to be explicit about what to do when it is null or a number so every case is taken care of
    let current_votes_for_entry : Nat = switch votes_for_entry {
      case null 0;
      case (?Nat) Nat;
    };

    //once we have the number of votes, update the votes for the entry
    votes.put(entry, current_votes_for_entry + 1);

    //Return the number of votes as an array (so frontend can display it)
    Iter.toArray(votes.entries())
  };

  public func resetVotes() : async [(Text, Nat)] {
      votes.put("Motoko", 0);
      votes.put("Rust", 0);
      votes.put("TypeScript", 0);
      votes.put("Python", 0);
      Iter.toArray(votes.entries())
  };

};
```

> ### Deploying the dapp locally

Reviewer 2: I wonder if we should already have a section on "Testing your code/dapp" here or at a later stage.

Reviewer 2: @letmejustputthishere and maybe @rvanasa What do you think where testing should be introduced?

Reviewer 3: i personally would introduce it at a later level. i think at this point its enough to do "testing" by calling the canister via `dfx` or the `candid ui`

            </label><br>
            <label>
                <input type="radio" name="option" value="Python"> Python
            </label><br>
            <button type="submit">Vote</button>
        </form>
      </div>

      <!-- Poll results appear here-->
      <h2 id="results-title">Results</h2>
      <div id="results"></div>
    </div>
    <button id="reset">Reset Poll</button>

</body>
</html>
```

:::info
#### What does this code do?
The HTML code above is a simple form that provides the end user with options to select using radio buttons. There is nothing 'Web3' or specific to the IC about this code.

Then, the `<head>` tag is used to include some basic CSS for the page's styling. To learn more about adding a stylesheet, see: [add a stylesheet](/docs/developer-docs/frontend/add-stylesheet.md).
:::

Lastly, we need to update the `index.js` file in order for our frontend canister to communicate with our backend canister and have the result of that communication reflected within the frontend's HTML. 

Open the `/src/poll_frontend/src/index.js` file in your code editor and replace the content of `index.js` with the following:

> ```javascript

Reviewer: as a general comment - I do think that in this day and age, vanilla JS may be more difficult to parse for new developers than say, React. many developers are introduced to the world of programming through tools like React; so much so that vanilla JS may appear daunting.

of course the trade off is that if you used a React app as the basis of the tutorial, you either have to explain the addition of all the libraries or hand-wave through all that stuff.

    e.preventDefault();
    
    //Reset the options in the backend
    await poll_backend.resetVotes();
    const voteCounts = await poll_backend.getVotes();
    updateLocalVoteCounts(voteCounts);

    //re-render the results once the votes are reset in the backend
    displayResults();
    return false;
}, false);

//3. HELPER FUNCTIONS

//Helper vanilla JS function to create the HTML to render the results of the poll
function displayResults() {
  let resultHTML = '<ul>';
  for (let key in pollResults) {
      resultHTML += '<li><strong>' + key + '</strong>: ' + pollResults[key] + '</li>';
  }
  resultHTML += '</ul>';
  resultsDiv.innerHTML = resultHTML;
};

//This helper updates the local JS object that teh browser holds
// Example JSON that the frontend will get using the values above
  // [["Motoko","0"],["Python","0"],["Rust","0"],["TypeScript","0"]]
function updateLocalVoteCounts(arrayOfVoteArrays){

  for (let voteArray of arrayOfVoteArrays) {
    //Example voteArray -> ["Motoko","0"]
    let voteOption = voteArray[0];
    let voteCount = voteArray[1];
    pollResults[voteOption] = voteCount;
  }

};
```

:::info
#### What does this code do?

> On line 4, the statement `import { poll_backend } from "../../declarations/poll_backend";` is what allows the frontend to import an interface for the backend canister and seamlessly send it via messages using Candid. This line is directly related to the following lines where the frontend JS talks to the backend:

Reviewer 2: ```suggestion
On line 4, the statement `import { poll_backend } from "../../declarations/poll_backend";` is what allows the frontend to import an interface for the backend canister and seamlessly interact with it. This line is directly related to the following lines where the frontend JS talks to the backend:
```

Reviewer 2: In general, I think this is a bit overwhelming for newcomers (both web2 and web3).

It might be better to start with the core js snippet on how to call a canister method from JS and explain it and then provide the complete code afterward.

  - Line 19: `const question = await poll_backend.getQuestion();`
  - Line 25: `const voteCounts = await poll_backend.getVotes();`
  - Line 40: `const updatedVoteCounts = await poll_backend.vote(selectedOption);`

`displayResults()` and `updateLocalVoteCount()` are helper functions created for convenience.

These lines collectively show how the frontend can use the `getQuestion()`, `getVotes()`, `vote()` methods we created in earlier sections.
:::

## Re-deploying the dapp

Re-deploy the dapp locally with the command:

```bash
dfx deploy
```

Now, when the terminal's output displays the following, open the `poll_frontend` URL in a web browser:

```
  Frontend canister via browser
    poll_frontend: http://127.0.0.1:4943/?canisterId=avqkn-guaaa-aaaaa-qaaea-cai
  Backend canister via Candid interface:
    poll_backend: http://127.0.0.1:4943/?canisterId=asrmz-lmaaa-aaaaa-qaaeq-cai&id=by6od-j4aaa-aaaaa-qaadq-cai
```

Your finished dapp should be displayed in the browser and look like this:

![Poll dapp](../_attachments/poll-dapp.png)

> Congratulations! You've just developed your first full dapp!

Reviewer 2: We should also provide a repo that they can just clone/fork.

## Next steps

> In this tutorial, we only deployed our dapp to our local environment. In the next module, we'll set up a cycles wallet and acquire cycles in order for us to deploy our dapp to the Internet Computer mainnet network. 

Reviewer 2: ```suggestion
In this tutorial, we only deployed our dapp to our local environment. In the next module, we'll set up a cycles wallet and acquire cycles in order for us to deploy our dapp to the Internet Computer mainnet. 
```
